File: test-any/go/read.go
package main

import (
	"fmt"
	"os"

	"google.golang.org/protobuf/proto"
)

func read(fname string) {
	data, err := os.ReadFile(fname)
	check(err)

	var task Task

	err = proto.Unmarshal(data, &task)
	check(err)

	fmt.Println(task.DoneBy)

	switch task.DoneBy.TypeUrl {
	case "type.googleapis.com/tutorial.Person":
		var r Person
		err = task.DoneBy.UnmarshalTo(&r)
		check(err)
		fmt.Println("age:", r.Age)

	case "type.googleapis.com/tutorial.Robot":
		var r Robot
		err = task.DoneBy.UnmarshalTo(&r)
		check(err)
		fmt.Println("features:", r.Features)

	default:
		fmt.Println("unknown type")
	}

}

File: test-any/py/read.py
import schema_pb2

def decode(fname):
  data = None
  with open(fname, "rb") as f:
    data = f.read()

  assert data is not None
  task1 = schema_pb2.Task()
  task1.ParseFromString(data)
  print(task1)
  typeName = task1.done_by.TypeName()
  print("\nby type:",typeName)

  print("\ndone_by:")
  if typeName == "tutorial.Person":
    person = schema_pb2.Person()
    task1.done_by.Unpack(person)
    print(person)
  elif typeName == "tutorial.Robot":
    robot = schema_pb2.Robot()
    task1.done_by.Unpack(robot)
    print(robot)
  else:
    print(f"unknown type {typeName}")

decode("task1.bin")
decode("task2.bin")
